pylint/plugins/hass_enforce_type_hints.py

```diff
@@ -0,0 +1,215 @@
+            return True
+        if isinstance(node, astroid.Attribute) and node.attrname == self.expected_type:
+            # Attribute occurs when a namespace is used, eg. "core.HomeAssistant"
+            return True
+        return False
+
+
+_MODULE_FILTERS: dict[str, re.Pattern] = {
+    "init": re.compile(r"^homeassistant.components.[a-zA-Z_]+$"),
+}
+
+_ARGUMENT_MATCH: dict[str, TypeChecker] = {
+    "HomeAssistant": TypeChecker("HomeAssistant"),
+    "ConfigType": TypeChecker("ConfigType"),
+    "ConfigEntry": TypeChecker("ConfigEntry"),
+}
+_RETURN_MATCH: dict[str, TypeChecker] = {
+    "bool": TypeChecker("bool"),
+    "None": TypeChecker(None),
+}
+
+_METHOD_MATCH: list[TypeHintMatch] = [
+    TypeHintMatch(
+        module_filter=_MODULE_FILTERS["init"],
+        function_name="setup",
+        arg_types={
+            0: _ARGUMENT_MATCH["HomeAssistant"],
+            1: _ARGUMENT_MATCH["ConfigType"],
+        },
+        return_type=_RETURN_MATCH["bool"],
+    ),
+    TypeHintMatch(
+        module_filter=_MODULE_FILTERS["init"],
+        function_name="async_setup",
+        arg_types={
+            0: _ARGUMENT_MATCH["HomeAssistant"],
+            1: _ARGUMENT_MATCH["ConfigType"],
+        },
+        return_type=_RETURN_MATCH["bool"],
+    ),
+    TypeHintMatch(
+        module_filter=_MODULE_FILTERS["init"],
+        function_name="async_setup_entry",
+        arg_types={
+            0: _ARGUMENT_MATCH["HomeAssistant"],
+            1: _ARGUMENT_MATCH["ConfigEntry"],
+        },
+        return_type=_RETURN_MATCH["bool"],
+    ),
+    TypeHintMatch(
+        module_filter=_MODULE_FILTERS["init"],
+        function_name="async_remove_entry",
+        arg_types={
+            0: _ARGUMENT_MATCH["HomeAssistant"],
+            1: _ARGUMENT_MATCH["ConfigEntry"],
+        },
+        return_type=_RETURN_MATCH["None"],
+    ),
+    TypeHintMatch(
+        module_filter=_MODULE_FILTERS["init"],
+        function_name="async_unload_entry",
+        arg_types={
+            0: _ARGUMENT_MATCH["HomeAssistant"],
+            1: _ARGUMENT_MATCH["ConfigEntry"],
+        },
+        return_type=_RETURN_MATCH["bool"],
+    ),
+    TypeHintMatch(
+        module_filter=_MODULE_FILTERS["init"],
+        function_name="async_migrate_entry",
+        arg_types={
+            0: _ARGUMENT_MATCH["HomeAssistant"],
+            1: _ARGUMENT_MATCH["ConfigEntry"],
+        },
+        return_type=_RETURN_MATCH["bool"],
+    ),
+]
+
+
+def _get_all_annotation(node: astroid.FunctionDef) -> list[astroid.NodeNG | None]:
+    args = node.args
+    annotations = (
+        args.posonlyargs_annotations + args.annotations + args.kwonlyargs_annotations
+    )
+    if args.vararg is not None:
+        annotations.append(args.varargannotation)
+    if args.kwarg is not None:
+        annotations.append(args.kwargannotation)
+    return annotations
+
+
+def _has_valid_annotations(
+    annotations: list[astroid.NodeNG | None],
+) -> list[astroid.NodeNG | None]:
```

Comment:
Did you run mypy on it?
```suggestion
) -> bool:
```

Reply:
I have now!